### PR TITLE
WIP: adds interceptor to sign query params

### PR DESCRIPTION
**Pull request by gokulchandra**
Initial implementation for #52.

I considered both the approaches mentioned in the issue. 

### DSL on params
- Simple, straightforward
- It was at the request level
- Limited to just the query params

### Defining an interceptor
- Defined at the client level
- Potential to access other aspects of the Request for signing the request

Different services sign their requests differently eg: [amazon](https://docs.aws.amazon.com/general/latest/gr/sigv4-create-string-to-sign.html) vs [twitter](https://developer.twitter.com/en/docs/basics/authentication/guides/creating-a-signature)

@rybalkinsd But would it be more flexible to let user the user define their own interceptor? Thoughts? 

We can further refactor the implementation based on the design we arrive at.

### CODE REVIEW

**Comment by rybalkinsd** [comment]
@gokulchandra I'll do my best to give feedback today or tomorrow

**Comment by gokulchandra** [comment]
@rybalkinsd Any updates on this PR?

File: src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package io.github.rybalkinsd.kohttp.interceptors
+
+import okhttp3.HttpUrl
+import okhttp3.Interceptor
+import okhttp3.Request
+import okhttp3.Response
+
+class SigningInterceptor(private val parameterName: String, private val signer: HttpUrl.() -> String) : Interceptor {
+    override fun intercept(chain: Interceptor.Chain): Response {
+        val userRequest: Request = chain.request()
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt]: Would it be a bit more simple to use `with(chain.request) { }`?
And the same thoughts for builders

File: src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package io.github.rybalkinsd.kohttp.interceptors
+
+import okhttp3.HttpUrl
+import okhttp3.Interceptor
+import okhttp3.Request
+import okhttp3.Response
+
+class SigningInterceptor(private val parameterName: String, private val signer: HttpUrl.() -> String) : Interceptor {
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt]: Need to refer common pattern:
LogInteceptor or LoggingInterceptor
Sign or Signing

Let’s define it in #64

Comment (rybalkinsd) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt]: Can we provide any default for param name?

Comment (rybalkinsd) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt]: Also need to add javadoc. We should mention if we guarantee any param order or not

File: src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt (hunk 1 above)
Comment (gokulchandra) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt]: Agreed, don't know how I missed that one! 😅

File: src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt (hunk 2 above)
Comment (gokulchandra) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt]: > Can we provide any default for param name?

Not sure. From what I saw all API's have different parameter names.

> Also need to add javadoc. We should mention if we guarantee any param order or not

Users have access to the `Url` object in the signer method, so I guess they have control over ordering.

File: src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ * @param signer function with HttpUrl as a receiver to sign the request parameter
+ *
+ * @see HttpUrl
+ *
+ * @since 0.8.0
+ * @author gokul
+ */
+
+class SigningInterceptor(private val parameterName: String, private val signer: HttpUrl.() -> String) : Interceptor {
+    override fun intercept(chain: Interceptor.Chain): Response {
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt]: looks much better 👍 
also it's possible to replace` } return ` with `=`
```suggestion
    override fun intercept(chain: Interceptor.Chain): Response =
```

File: src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+        return with(chain.request()) {
+            val signedKey = signer(url())
+
+            val requestUrl = with(url().newBuilder()) {
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt]: better to add a long term issue for url DSL. don't have a manual url building in our dsl now, but I think will have in a near future.

File: src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt (hunk 2 above)
Comment (rybalkinsd) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt]: Yeah, looks like internal logic is straightforward. They use `ArrayList` and `add` method, like append to the end of list. Better to mention that it works this way in our class docs.

see `okhttp3.HttpUrl.Builder#addQueryParameter`

File: src/test/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptorTest.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+class SigningInterceptorTest {
+    @Test
+    fun `signs request with MD5 algorithm`() {
+        val urlEncoder = Base64.getUrlEncoder()
+        val algorithm = MessageDigest.getInstance("md5")
+
+        val client = defaultHttpClient.fork {
+            interceptors = listOf(SigningInterceptor("key") {
+                val query = (query() ?: "").toByteArray()
```
Comment (rybalkinsd) [inline src/test/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptorTest.kt]: This peace of code is most related to user experience. It should be as easy to interact with as possible.
 
Possible suggestions here:
- We can introduce extension function for `MessageDigest` to encode `String?` in `MessageDigestExt` file
- It's also possible to provide a util function like `md5(data: String?)`

From my experience it's also may be useful to provide ext to operate with `Map<String, List<String>` (map of param names to their values.

Comment (rybalkinsd) [reply src/test/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptorTest.kt]: I'll check with android buddies what they are using tomorrow :)

File: src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt (hunk 4 above)
Comment (gokulchandra) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt]: URL builder DSL: #71

**Comment by rybalkinsd** [comment]
Will create release/0.8.0 branch and start merging to it

File: src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * @see HttpUrl.Builder.addQueryParameter
+ *
+ * @since 0.8.0
+ * @author gokul
+ */
+
+class SigningInterceptor(private val parameterName: String, private val signer: HttpUrl.() -> String) : Interceptor {
+    override fun intercept(chain: Interceptor.Chain): Response =
+
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt]: m/b omit this line, or event put `with(chain.request()) {` right after `: Response = ` ?

File: src/test/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptorTest.kt (hunk 5 above)
Comment (rybalkinsd) [reply src/test/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptorTest.kt]: > Do we provide this on the Interceptor?
three options here:
1. change `signer` `HttpUrl.() -> String` to `Map<String, List<String>> -> String`
2. support both
3. support both + `String -> String` ("foo=bar&random=213" -> hash(...))

Comment (rybalkinsd) [reply src/test/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptorTest.kt]: Let's have ext only for `hmac` and `md5` now.

**Comment by rybalkinsd** [comment]
@gokulchandra could u pls change target branch to release/0.8.0 ?

File: src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt (hunk 6 above)
Comment (gokulchandra) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptor.kt]: Yep. I'll update it.

**Comment by rybalkinsd** [comment]
Great job @gokulchandra !

File: src/test/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptorTest.kt (hunk 5 above)
Comment (gokulchandra) [reply src/test/kotlin/io/github/rybalkinsd/kohttp/interceptors/SigningInterceptorTest.kt]: > > Do we provide this on the Interceptor?
> > three options here:
> 
> 1. change `signer` `HttpUrl.() -> String` to `Map<String, List<String>> -> String`
> 2. support both
> 3. support both + `String -> String` ("foo=bar&random=213" -> hash(...))

> Let's have ext only for `hmac` and `md5` now.

I haven't had a chance to work on these yet, I will have a think about design and raise a PR for it.

**Pull request merged**